src/AppBundle/EventListener/OccurrenceListener.php

```diff
@@ -0,0 +1,149 @@
+{
+    private $occurrenceSplitter;
+
+    /**
+     * OccurrenceListener constructor.
+     *
+     * @param OccurrenceSplitterService $occurrenceSplitter
+     */
+    public function __construct(OccurrenceSplitterService $occurrenceSplitter)
+    {
+        $this->occurrenceSplitter = $occurrenceSplitter;
+    }
+
+    /**
+     * @param PreFlushEventArgs $args
+     *
+     * @throws \Exception
+     */
+    public function onFlush(OnFlushEventArgs $args): void
+    {
+        $em = $args->getEntityManager();
+
+        $this->insertDailyOccurrences($em);
+        $this->deleteDailyOccurrences($em);
+        $this->synchronizeDailyOccurrences($em);
+    }
+
+    /**
+     * Insert new DailyOccurrences matching the Occurrences scheduled for insertion
+     * in doctrines unit of work.
+     *
+     * @param EntityManager $em
+     *
+     * @throws \Doctrine\ORM\ORMException
+     */
+    private function insertDailyOccurrences(EntityManager $em): void
+    {
+        $uow = $em->getUnitOfWork();
+        $entities = $uow->getScheduledEntityInsertions();
+
+        $occurrences = array_filter($entities, function ($entity) {
+            return $entity instanceof Occurrence;
+        });
+
+        $classMetadata = $em->getClassMetadata(DailyOccurrence::class);
+
+        foreach ($occurrences as $occurrence) {
+            $dailyOccurrences = $this->occurrenceSplitter->getDailyOccurrences($occurrence);
+
+            foreach ($dailyOccurrences as $dailyOccurrence) {
+                $em->persist($dailyOccurrence);
+                $uow->computeChangeSet($classMetadata, $dailyOccurrence);
+            }
+        }
+    }
+
+    /**
+     * Synchronize the DailyOccurrences to match the Occurrences scheduled for updates
+     * in doctrines unit of work.
+     *
+     * @param EntityManager $em
+     *
+     * @throws \Doctrine\ORM\ORMException
+     */
+    private function synchronizeDailyOccurrences(EntityManager $em): void
+    {
+        $uow = $em->getUnitOfWork();
+        $entities = $uow->getScheduledEntityUpdates();
+
+        $occurrences = array_filter($entities, function ($entity) {
+            return $entity instanceof Occurrence;
+        });
+
+        $classMetadata = $em->getClassMetadata(DailyOccurrence::class);
+
+        foreach ($occurrences as $occurrence) {
+            $newDailyOccurrences = $this->occurrenceSplitter->getDailyOccurrences($occurrence);
+            $existingDailyOccurrences = $em->getRepository(DailyOccurrence::class)->findByOccurrence($occurrence);
+
+            // Loop through new DailyOccurrences and copy their data to the first exiting DailyOccurrence to update
+            // instead of doing delete/insert. Then remove the exiting DailyOccurrence.
+            // Once there are no more exiting DailyOccurrences we persist new entities.
+            $count = 0;
+            $totalExisting = \count($existingDailyOccurrences);
+            foreach ($newDailyOccurrences as $newDailyOccurrence) {
```

Comment:
Reusing occurrences is a good idea and the goal is to not change `id`s of existing occurrences (I suppose). However, unless we know (or assume) something about the order of occurrences we risk actually changing the `id`s anyway, e.g.

```
[
{ "startTime": '2001-01-01', 'endTime': '2002-01-01' },
{ "startTime": '2011-01-01', 'endTime': '2012-01-01' }
]
```

could become

```
[
{ "startTime": '2011-01-01', 'endTime': '2012-01-01' },
{ "startTime": '2001-01-01', 'endTime': '2002-01-01' }
]
```

This is a minor issue – and probably not a real issue at all …

Reply:
We know that the order of `DailyOccurrences` is always chronological because they are always created and updated by our service. 

I considered a number of strategies for determining if a `DailyOccurrence` existed already. I rejected all of them because of complexity and decreased performance on large collections when doing the "exists in collection" check. 

The current implementation is not perfect. There are cases where it gives a sub-optimal result. Fx if the `startDate` of an `Occurrence` is moved forward by one day. Then all  `DailyOccurrences` will shift forward one day by id as opposed to just inserting one extra `DailyOccurrence`. But given the above considerations, and the fact that `/daily_occurrences` is not writable i settled on this compromise.